src/main/java/com/redhat/devtools/lsp4ij/features/codeBlockProvider/LSPCodeBlockProvider.java

```diff
@@ -0,0 +1,201 @@
+        // offset for selection ranges to act as if we're in the adjusted braced block
+        TextRange codeBlockRange = getUsingSelectionRanges(
+                file,
+                editor,
+                offset,
+                openBraceChar,
+                openBraceOffset,
+                closeBraceChar,
+                closeBraceOffset
+        );
+        if (codeBlockRange != null) {
+            return codeBlockRange;
+        }
+
+        // Failing that, try to find it using the folding ranges
+        return getUsingFoldingRanges(
+                file,
+                document,
+                offset,
+                openBraceChar,
+                closeBraceChar
+        );
+    }
+
+    @Nullable
+    private TextRange getUsingSelectionRanges(@NotNull PsiFile file,
+                                              @NotNull Editor editor,
+                                              int offset,
+                                              @Nullable Character openBraceChar,
+                                              int openBraceOffset,
+                                              @Nullable Character closeBraceChar,
+                                              int closeBraceOffset) {
+        Document document = editor.getDocument();
+        List<SelectionRange> selectionRanges = LSPSelectionRangeSupport.getSelectionRanges(file, document, offset);
+        if (!ContainerUtil.isEmpty(selectionRanges)) {
+            // Convert the selection ranges into text ranges
+            Set<TextRange> textRanges = new LinkedHashSet<>(selectionRanges.size());
+            for (SelectionRange selectionRange : selectionRanges) {
+                textRanges.add(LSPSelectionRangeSupport.getTextRange(selectionRange, document));
+                for (SelectionRange parentSelectionRange = selectionRange.getParent();
+                     parentSelectionRange != null;
+                     parentSelectionRange = parentSelectionRange.getParent()) {
+                    textRanges.add(LSPSelectionRangeSupport.getTextRange(parentSelectionRange, document));
+                }
+            }
+
+            // Find containing text ranges that are bounded by brace pairs
+            CharSequence documentChars = document.getCharsSequence();
+            List<TextRange> containingTextRanges = new ArrayList<>(textRanges.size());
+            for (TextRange textRange : textRanges) {
+                if (textRange.getLength() > 1) {
+                    int startOffset = textRange.getStartOffset();
+                    int endOffset = textRange.getEndOffset();
+
+                    char startChar = documentChars.charAt(startOffset);
+                    char endChar = documentChars.charAt(endOffset - 1);
+
+                    // If aligned on an open brace and this ends with the expected close brace, use it
+                    if ((startOffset == openBraceOffset)) {
```

Comment:
The first two conditions here are for when the caret starts exactly on an open or close brace(/bracket/paren) and the evaluated range begins/ends on that offset and ends/begins with the appropriate paired brace. Basically these are an exact brace match.